src/core/classes/language.ts

```diff
@@ -0,0 +1,78 @@
+import { extend, insertBefore } from '../../shared/language-util';
+import type LanguageRegistry from './language-registry';
+import type { LanguageProto, Grammar } from '../../types';
+
+export default class Language {
+	def: LanguageProto;
+	registry: LanguageRegistry;
+	evaluatedGrammar?: Grammar;
+
+	constructor (def: LanguageProto, registry: LanguageRegistry) {
+		this.def = def;
+		this.registry = registry;
+	}
+
+	get id () {
+		return this.def.id;
+	}
+
+	get alias () {
+		if (!this.def.alias) {
+			return [];
+		}
+
+		return Array.isArray(this.def.alias) ? this.def.alias : [this.def.alias];
+	}
+
+	get grammar (): Grammar {
+		if (!this.evaluatedGrammar) {
+			// Evaluate grammar
+			const def = this.def;
+			const grammar = this.def.grammar;
+
+			// TODO extends should also be able to
+			let baseGrammar = def.extends ? { ...def.extends } : {};
+
+			if (typeof grammar === 'object') {
+				// the grammar is a simple object, so we don't need to evaluate it
+				this.evaluatedGrammar = Object.assign(baseGrammar, grammar);
+			}
+			else if (typeof grammar === 'function') {
+				this.evaluatedGrammar = grammar.call(this);
+			}
```

Comment:
We use `else if` here, which means that `this.evaluatedGrammar` might be `undefined` after this block. It `undefined` is a possible value, we should fix the returned type to `Grammar | undefined`.

Do I get it right that if the `grammar` getter returns `undefined`, something is wrong?